### ISSUE TITLE
 Bad error message when closure parameter type can't be inferred

### ISSUE THREAD

**Issue by rbtcollins** [rust-lang/rust#71732]
This is either a regression of #45685 or it was closed prematurely.

Fairly minimal reproducer:
```rust
use std::collections::hash_map::HashMap;

fn foo(parameters: &HashMap<String, String>) -> bool {
    parameters
        .get(&"key".into())
        .and_then(|found: &String| {
            Some(false)
        })
        .unwrap_or(false)
}
```
->
```
error[E0283]: type annotations needed for `&std::string::String`
 --> src/lib.rs:6:20
  |
6 |         .and_then(|found: &String| {
  |                    ^^^^^ consider giving this closure parameter the explicit type `&std::string::String`, where the type parameter `std::string::String` is specified
  |
  = note: cannot resolve `std::string::String: std::borrow::Borrow<_>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0283`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c8754ceba0148a4da807002ae99ffa25

I'm still not entirely sure what is actually wrong here - is it the lifetime as in #22340? Anyhow - the message doesn't refer to any T that needs to be specified, and the named explicit type is not sufficient to fix the issue even when given.

Happens with nightly too.

**Comment by rbtcollins** [comment]
Right, so the triggering problem here is the `get()` call - with that fixed it all falls into place, but the error message is still poor.

**Comment by adaszko** [comment]
> Right, so the triggering problem here is the `get()` call - with that fixed it all falls into place, but the error message is still poor.

I got hit by this bug too and in my case the fix isn't obvious.  What do meant by "fixing the get() call"?

**Comment by rbtcollins** [comment]
It has a parameter type that is a bit bonkers given the type of the Hash, which causes the type inference of the `into()` to go off into the weeds I think; doesn't change that there is a bug - this is about the diagnostics after all.

```
use std::collections::hash_map::HashMap;

fn foo(parameters: &HashMap<String, String>) -> bool {
    parameters
        .get("key")
        .and_then(|_found: &String| {
            Some(false)
        }).unwrap_or(false)
}
```

is a fixed version of the broken code.



https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b4332041d4727de8e53025cf2aee89cb

**Comment by lcnr** [comment]
The error message should have been fixed by #78111, marking as `E-needs-test`.
